rskj-core/src/main/java/co/rsk/peg/whitelist/WhitelistSupportImpl.java

```diff
@@ -0,0 +1,226 @@
+package co.rsk.peg.whitelist;
+
+import static co.rsk.peg.whitelist.WhitelistResponseCode.ADDRESS_ALREADY_WHITELISTED;
+import static co.rsk.peg.whitelist.WhitelistResponseCode.ADDRESS_NOT_EXIST;
+import static co.rsk.peg.whitelist.WhitelistResponseCode.DELAY_ALREADY_SET;
+import static co.rsk.peg.whitelist.WhitelistResponseCode.DISABLE_BLOCK_DELAY_INVALID;
+import static co.rsk.peg.whitelist.WhitelistResponseCode.INVALID_ADDRESS_FORMAT;
+import static co.rsk.peg.whitelist.WhitelistResponseCode.SUCCESS;
+import static co.rsk.peg.whitelist.WhitelistResponseCode.UNAUTHORIZED_CALLER;
+import static co.rsk.peg.whitelist.WhitelistResponseCode.UNKNOWN_ERROR;
+
+import co.rsk.bitcoinj.core.Address;
+import co.rsk.bitcoinj.core.AddressFormatException;
+import co.rsk.bitcoinj.core.Coin;
+import co.rsk.bitcoinj.core.Context;
+import co.rsk.bitcoinj.store.BlockStoreException;
+import co.rsk.peg.vote.AddressBasedAuthorizer;
+import co.rsk.peg.whitelist.constants.WhitelistConstants;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.util.List;
+import org.ethereum.core.SignatureCache;
+import org.ethereum.core.Transaction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of the WhitelistSupport interface. Contains all the whitelist business logic.
+ */
+public class WhitelistSupportImpl implements WhitelistSupport {
+
+    private static final Logger logger = LoggerFactory.getLogger(WhitelistSupportImpl.class);
+    private static final String INVALID_ADDRESS_FORMAT_MESSAGE = "invalid address format";
+    private final SignatureCache signatureCache;
+    private final Context btcContext;
+    private final WhitelistConstants whitelistConstants;
+    private final WhitelistStorageProvider whitelistStorageProvider;
+
+    public WhitelistSupportImpl(
+        WhitelistStorageProvider whitelistStorageProvider,
+        WhitelistConstants whitelistConstants,
+        SignatureCache signatureCache,
+        Context btcContext) {
+        this.whitelistStorageProvider = whitelistStorageProvider;
+        this.whitelistConstants = whitelistConstants;
+        this.signatureCache = signatureCache;
+        this.btcContext = btcContext;
+    }
+
+    /**
+     * Returns the lock whitelist size, that is, the number of whitelisted addresses
+     *
+     * @return the lock whitelist size
+     */
+    @Override
+    public int getLockWhitelistSize() {
+        return whitelistStorageProvider.getLockWhitelist().getSize();
+    }
+
+    /**
+     * Returns the lock whitelist address stored at the given index, or null if the index is out of
+     * bounds
+     *
+     * @param index the index at which to get the address
+     * @return the base58-encoded address stored at the given index, or null if index is out of
+     * bounds
+     */
+    @Override
+    public LockWhitelistEntry getLockWhitelistEntryByIndex(int index) {
```

Comment:
What about returning `Optional<LockWhitelistEntry>` 🤔. This way we avoid returning and managing `null`. 
What do you guyst think? @wilmerrootstock , @apancorb , @nathanieliov , @marcos-iov .

Reply:
That's the plan yes, we have a separate task to tackle that

Reply:
@jeremy-then, Totally agree! @apancorb also recommended the same in a previous PR. It's a good improvement!
We'll address this improvement in another task as @marcos-iov brought up.